Objects/memoryobject.c

```diff
@@ -3160,6 +3162,133 @@ static PyMethodDef memory_methods[] = {
     {NULL,          NULL}
 };
 
+/**************************************************************************/
+/*                          Memoryview Iterator                           */
+/**************************************************************************/
+
+typedef struct {
+    PyObject_HEAD
+    Py_ssize_t it_index;
+    PyMemoryViewObject *it_seq; // Set to NULL when iterator is exhausted
+    Py_ssize_t length;
+    const char *fmt;
+} memoryiterobject;
+
+
+static void memoryiter_dealloc(memoryiterobject *);
+static int memoryiter_traverse(memoryiterobject * , visitproc, void *);
+static PyObject *memoryiter_next(memoryiterobject *);
+static PyObject *memory_iter(PyObject *);
+
+
+static PyTypeObject PyMemoryIter_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "memory_iterator",
+    .tp_basicsize = sizeof(memoryiterobject),
+    // methods
+    .tp_dealloc = (destructor)memoryiter_dealloc,
+    .tp_getattro = PyObject_GenericGetAttr,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_traverse = (traverseproc)memoryiter_traverse,
+    .tp_iter = PyObject_SelfIter,
+    .tp_iternext = (iternextfunc)memoryiter_next,
+};
+
+static void
+memoryiter_dealloc(memoryiterobject *it)
+{
+    _PyObject_GC_UNTRACK(it);
+    Py_XDECREF(it->it_seq);
+    PyObject_GC_Del(it);
+}
+
+static int
+memoryiter_traverse(memoryiterobject *it, visitproc visit, void *arg)
+{
+    Py_VISIT(it->it_seq);
+    return 0;
+}
+
+static char *
```

Comment:
Consider making this function `inline`.

Reply:
It's not faster than bytes for me yet, but it's definitely the fastest I've seen it. On my platform the time went from 12.7 to 9.96 msec (looping over a memoryview 1,000,000 of bytes.) Congratulations!

The main overhead that `--with-pydebug` adds is not so much compiler optimizations as extra asserts. E.g. remember the assert you added to your `GET_SIZE` macro? Probably the memoryview code still has more asserts than the bytes iterator.

A fun experiment: if the fmt code is `'B'` and the data is contiguous, inline even more (you can just call `PyLong_FromLong`, see `unpack_single`.

Reply:
All credit goes to you. Thank you once again for all the time you are taking to teach me, I can't wait to learn more from you and I couldn't be more grateful.

I will try your idea after dinner and report on it. On that note, wouldn't we be able to use that same hack for all the different formats, if we could infer that the data is contiguous?

Reply:
Well, you don't want to duplicate all the logic from `unpack_single()`. In the end, who's going to care whether iterating over a memoryview takes 10 nsec or 11 nsec per iteration. For yourself, what's the most important data type when you're iterating over a numpy array?

Reply:
Float.

I tried the idea you mentioned above and my specific implementation (which might be non-optimal) made things significantly slower.

Back to the current commit, when compiling with optimizations enabled, iterating over bytes is still ~30% faster than iterating over memoryview on my setup, however
with debug enabled, memoryview is ~1% faster.

Do you see any further optimizations, besides the one you mentioned, that could be done to try and close the difference between bytes and memoryview?

Reply:
No, I think we should leave well enough alone. This is only a micro-optimization anyway -- nobody's going to care about any further optimizations, but the core devs would object against the added complexity of the code (e.g. if someone were to change the implementation of `ptr_from_index` or `lookup_dimension` they would have to remember to also update `memoryiter_next`.

i have one further suggestion before we move on to bigger and better things. If you move `PyMemoryIter_Type` after the various functions you won't need to add those forward declarations. This is how things are traditionally written. (For example, this is why `PyMemoryView_Type` is at the end of the file.)

Reply:
Hi Guido,

I pushed a new commit that addresses your suggestion above, however after some consideration I opted to forward declare PyMemoryIter_type (instead of all the functions) because otherwise I would have to leave it above memor_iter() - which uses it to register with the garbage collector - and in my opinion it would negatively affect the reading flow. 
I think the code is very readable this way, but please let me know if you would prefer no forward declarations at all.

> (...) before we move on to bigger and better things.

Can't wait.